src/core/geometry/qgsgeometryengine.cpp

```diff
@@ -0,0 +1,90 @@
+/***************************************************************************
+                        qgsgeometryengine.cpp
+  -------------------------------------------------------------------
+Date                 : 23.12.2015
+Copyright            : (C) 2014 by Matthias Kuhn
+email                : matthias@opengis.ch
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsgeometryengine.h"
+
+#include "qgsmultipolygonv2.h"
+#include "qgsmultilinestringv2.h"
+#include "qgspolygonv2.h"
+
+#include <QTransform>
+
+QgsGeometry QgsGeometryEngine::explode()
+{
+  QgsGeometry result;
+  switch ( mGeometry->wkbType() )
```

Comment:
this should be `switch ( QgsWKBTypes::flatType( mGeometry->wkbType() ) )`, otherwise it won't handle MultiLineStringZ/M/ZM/etc...